### DISCUSSION TITLE
unable to clone private git-submodules

### DISCUSSION THREAD

**Discussion by ab77** [renovatebot/renovate#27280] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.187.1

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

I have not seen this working (although [this](https://github.com/renovatebot/renovate/issues/10149) supposedly fixed it)

### Describe the problem

> renovating a private repo containing a private git-submodule fails with `fatal: could not read Username for 'https://github.com/': No such device or address` error

Minimal reproduction setup:
- [renovate-config](https://github.com/belodetek/renovate-config) (public)
- `https://github.com/belodetek/minimal-reproduction` (private)
- `https://github.com/belodetek/private-submodule` (private)

```
minimal-reproduction
├── README.md
└── private-submodule (https://github.com/belodetek/private-submodule.git (push,fetch)
    └── README.md

private-submodule
└── README.md
```

With the above configuration, we are not able to get Renovate to clone `private-submodule` when renovating `minimal-reproduction` repo. However, making `private-submodule` repository public, allows Renovate to clone `minimal-reproduction` repo with submodules, as expected. As long as `private-submodule` repo is private, Renovate throws:

<details><summary>Git function thrown</summary>

```
DEBUG: Cloning git submodule at private-submodule (repository=belodetek/minimal-reproduction)
DEBUG: Git function thrown (repository=belodetek/minimal-reproduction)
       "err": {
         "task": {
           "commands": ["submodule", "update", "--init", "private-submodule"],
           "format": "utf-8",
           "parser": "[function]"
         },
         "message": "Submodule 'private-submodule' (https://github.com/belodetek/private-submodule.git) registered for path 'private-submodule'\nCloning into '/tmp/renovate/repos/github/belodetek/minimal-reproduction/private-submodule'...\nfatal: could not read Username for 'https://github.com': No such device or address\nfatal: clone of 'https://github.com/belodetek/private-submodule.git' into submodule path '/tmp/renovate/repos/github/belodetek/minimal-reproduction/private-submodule' failed\nFailed to clone 'private-submodule'. Retry scheduled...
```

</details>

Full Renovate log [here](https://github.com/belodetek/renovate-config/actions/runs/7893986374/job/21543723023).

### Relevant debug logs

<details><summary>Logs</summary>

```
...
DEBUG: File config
       "config": {
         "extends": [
           "config:recommended",
           "config:best-practices",
           "helpers:pinGitHubActionDigests",
           "regexManagers:githubActionsVersions"
         ],
         "cloneSubmodules": true,
         "onboarding": false,
         "printConfig": true,
         "requireConfig": "optional",
         "git-submodules": {"enabled": true},
         "repositories": ["belodetek/minimal-reproduction"]
       }
DEBUG: CLI config
       "config": {}
DEBUG: Env config
       "config": {
         "hostRules": [
           {
             "hostType": "github",
             "matchHost": "api.github.com",
             "token": "***********"
           },
           {
             "hostType": "git-refs",
             "matchHost": "api.github.com",
             "token": "***********"
           }
         ],
         "dryRun": "full",
         "token": "***********"
       }
DEBUG: Combined config
       "config": {
         "extends": [
           "config:recommended",
           "config:best-practices",
           "helpers:pinGitHubActionDigests",
           "regexManagers:githubActionsVersions"
         ],
         "cloneSubmodules": true,
         "onboarding": false,
         "printConfig": true,
         "requireConfig": "optional",
         "git-submodules": {"enabled": true},
         "repositories": ["belodetek/minimal-reproduction"],
         "hostRules": [
           {
             "hostType": "github",
             "matchHost": "api.github.com",
             "token": "***********"
           },
           {
             "hostType": "git-refs",
             "matchHost": "api.github.com",
             "token": "***********"
           }
         ],
         "dryRun": "full",
         "token": "***********"
       }
DEBUG: Enabling forkProcessing while in non-autodiscover mode
DEBUG: Found valid git version: 2.43.0
DEBUG: Setting global hostRules
DEBUG: Adding token authentication for api.github.com (hostType=github) to hostRules
DEBUG: Adding token authentication for api.github.com (hostType=git-refs) to hostRules
DEBUG: Using default github endpoint: https://api.github.com/
DEBUG: hostRules: authentication already set for api.github.com
DEBUG: Platform config
       "platformConfig": {
         "hostType": "github",
         "endpoint": "https://api.github.com/",
         "isGHApp": true,
         "isGhe": false,
         "userDetails": {
           "username": "belodetek[bot]",
           "name": "belodetek[bot]",
           "id": 124739940
         }
       },
       "renovateUsername": "belodetek[bot]"
DEBUG: Using platform gitAuthor: belodetek[bot] <124739940+belodetek[bot]@users.noreply.github.com>
DEBUG: Adding token authentication for api.github.com (hostType=github) to hostRules
DEBUG: Using baseDir: /tmp/renovate
DEBUG: Using cacheDir: /tmp/renovate/cache
DEBUG: Using containerbaseDir: /tmp/renovate/cache/containerbase
DEBUG: Initializing Renovate internal cache into /tmp/renovate/cache/renovate/renovate-cache-v1
DEBUG: Commits limit = null
DEBUG: Setting global hostRules
DEBUG: Adding token authentication for api.github.com (hostType=github) to hostRules
DEBUG: Adding token authentication for api.github.com (hostType=git-refs) to hostRules
DEBUG: Adding token authentication for api.github.com (hostType=github) to hostRules
DEBUG: validatePresets()
DEBUG: Reinitializing hostRules for repo
DEBUG: Clearing hostRules
DEBUG: Adding token authentication for api.github.com (hostType=github) to hostRules
DEBUG: Adding token authentication for api.github.com (hostType=git-refs) to hostRules
DEBUG: Adding token authentication for api.github.com (hostType=github) to hostRules
 INFO: Repository started (repository=belodetek/minimal-reproduction)
       "renovateVersion": "37.187.1"
DEBUG: Using localDir: /tmp/renovate/repos/github/belodetek/minimal-reproduction (repository=belodetek/minimal-reproduction)
DEBUG: PackageFiles.clear() - Package files deleted (repository=belodetek/minimal-reproduction)
DEBUG: initRepo("belodetek/minimal-reproduction") (repository=belodetek/minimal-reproduction)
DEBUG: hostRules: authentication already set for api.github.com (repository=belodetek/minimal-reproduction)
DEBUG: belodetek/minimal-reproduction default branch = master (repository=belodetek/minimal-reproduction)
DEBUG: Using app token for git init (repository=belodetek/minimal-reproduction)
DEBUG: Resetting npmrc (repository=belodetek/minimal-reproduction)
DEBUG: Resetting npmrc (repository=belodetek/minimal-reproduction)
DEBUG: checkOnboarding() (repository=belodetek/minimal-reproduction)
DEBUG: isOnboarded() (repository=belodetek/minimal-reproduction)
DEBUG: Repo is onboarded (repository=belodetek/minimal-reproduction)
DEBUG: Initializing git repository into /tmp/renovate/repos/github/belodetek/minimal-reproduction (repository=belodetek/minimal-reproduction)
DEBUG: Performing blobless clone (repository=belodetek/minimal-reproduction)
DEBUG: git clone completed (repository=belodetek/minimal-reproduction)
       "durationMs": 1202
DEBUG: Cloning git submodule at private-submodule (repository=belodetek/minimal-reproduction)
DEBUG: Git function thrown (repository=belodetek/minimal-reproduction)
       "err": {
         "task": {
           "commands": ["submodule", "update", "--init", "private-submodule"],
           "format": "utf-8",
           "parser": "[function]"
         },
         "message": "Submodule 'private-submodule' (https://github.com/belodetek/private-submodule.git) registered for path 'private-submodule'\nCloning into '/tmp/renovate/repos/github/belodetek/minimal-reproduction/private-submodule'...\nfatal: could not read Username for 'https://github.com': No such device or address\nfatal: clone of 'https://github.com/belodetek/private-submodule.git' into submodule path '/tmp/renovate/repos/github/belodetek/minimal-reproduction/private-submodule' failed\nFailed to clone 'private-submodule'. Retry scheduled\nCloning into '/tmp/renovate/repos/github/belodetek/minimal-reproduction/private-submodule'...\nfatal: could not read Username for 'https://github.com': No such device or address\nfatal: clone of 'https://github.com/belodetek/private-submodule.git' into submodule path '/tmp/renovate/repos/github/belodetek/minimal-reproduction/private-submodule' failed\nFailed to clone 'private-submodule' a second time, aborting\n",
         "stack": "Error: Submodule 'private-submodule' (https://**redacted**@3.22.0/node_modules/simple-git/src/lib/plugins/error-detection.plugin.ts:42:29)\n    at PluginStore.exec (/usr/local/renovate/node_modules/.pnpm/simple-git@3.22.0/node_modules/simple-git/src/lib/plugins/plugin-store.ts:29:29)\n    at /usr/local/renovate/node_modules/.pnpm/simple-git@3.22.0/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:[127](https://github.com/belodetek/renovate-config/actions/runs/7893986374/job/21543723023#step:4:128):42\n    at new Promise (<anonymous>)\n    at GitExecutorChain.handleTaskData (/usr/local/renovate/node_modules/.pnpm/simple-git@3.22.0/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:124:14)\n    at GitExecutorChain.<anonymous> (/usr/local/renovate/node_modules/.pnpm/simple-git@3.22.0/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:100:40)\n    at Generator.next (<anonymous>)\n    at fulfilled (/usr/local/renovate/node_modules/.pnpm/simple-git@3.22.0/node_modules/simple-git/dist/cjs/index.js:55:24)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)"
       }
 WARN: Unable to initialise git submodule at private-submodule (repository=belodetek/minimal-reproduction)
       "err": {
         "task": {
           "commands": ["submodule", "update", "--init", "private-submodule"],
           "format": "utf-8",
           "parser": "[function]"
         },
         "message": "Submodule 'private-submodule' (https://github.com/belodetek/private-submodule.git) registered for path 'private-submodule'\nCloning into '/tmp/renovate/repos/github/belodetek/minimal-reproduction/private-submodule'...\nfatal: could not read Username for 'https://github.com/': No such device or address\nfatal: clone of 'https://github.com/belodetek/private-submodule.git' into submodule path '/tmp/renovate/repos/github/belodetek/minimal-reproduction/private-submodule' failed\nFailed to clone 'private-submodule'. Retry scheduled\nCloning into '/tmp/renovate/repos/github/belodetek/minimal-reproduction/private-submodule'...\nfatal: could not read Username for 'https://github.com/': No such device or address\nfatal: clone of 'https://github.com/belodetek/private-submodule.git' into submodule path '/tmp/renovate/repos/github/belodetek/minimal-reproduction/private-submodule' failed\nFailed to clone 'private-submodule' a second time, aborting\n",
         "stack": "Error: Submodule 'private-submodule' (https://**redacted**@3.22.0/node_modules/simple-git/src/lib/plugins/error-detection.plugin.ts:42:29)\n    at PluginStore.exec (/usr/local/renovate/node_modules/.pnpm/simple-git@3.22.0/node_modules/simple-git/src/lib/plugins/plugin-store.ts:29:29)\n    at /usr/local/renovate/node_modules/.pnpm/simple-git@3.22.0/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:127:42\n    at new Promise (<anonymous>)\n    at GitExecutorChain.handleTaskData (/usr/local/renovate/node_modules/.pnpm/simple-git@3.22.0/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:124:14)\n    at GitExecutorChain.<anonymous> (/usr/local/renovate/node_modules/.pnpm/simple-git@3.22.0/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:100:40)\n    at Generator.next (<anonymous>)\n    at fulfilled (/usr/local/renovate/node_modules/.pnpm/simple-git@3.22.0/node_modules/simple-git/dist/cjs/index.js:55:24)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)"
       }
...
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

### Additional information
> potentially related

* https://github.com/renovatebot/renovate/issues/16561
* https://github.com/renovatebot/renovate/issues/5471

**Comment by ab77** [comment, marked as answer]
This is a [test](https://github.com/belodetek/renovate-config/actions/runs/8193283966/job/22406680247#step:5:11) to confirm the token has access to private-submodule

[This](https://github.com/renovatebot/renovate/blob/37.227.0/lib/util/git/index.ts#L357) is where the exception is thrown.

.. and with debug/trace on, it's clear that [no auth. header is present](https://github.com/belodetek/renovate-config/actions/runs/8194433954/job/22410457067#step:4:1548) on initial submodule clone(s).

.. and it turns out this is [documented](https://docs.renovatebot.com/configuration-options/#clonesubmodules):

> Important: private submodules aren't supported by Renovate, unless the underlying ssh layer already has the correct permissions.

So, to summarise, `{ cloneSubmodules": true }` is not required in order to have Renovate update submodules, which is configured separately using `git-submodules` manager:

```
{
  "git-submodules": {
    "enabled": true
  }
}
```
